components/de_amp/browser/de_amp_throttle.h

```diff
@@ -0,0 +1,49 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_DE_AMP_BROWSER_DE_AMP_THROTTLE_H_
+#define BRAVE_COMPONENTS_DE_AMP_BROWSER_DE_AMP_THROTTLE_H_
+
+#include <memory>
+
+#include "base/memory/scoped_refptr.h"
+#include "base/memory/weak_ptr.h"
+#include "brave/components/de_amp/browser/de_amp_service.h"
+#include "brave/components/sniffer/sniffer_throttle.h"
+#include "content/public/browser/web_contents.h"
+#include "services/network/public/mojom/url_response_head.mojom-forward.h"
+
+class HostContentSettingsMap;
+
+namespace de_amp {
+
+// Throttle for AMP HTML detection.
+// If AMP page, redirect request to non-AMP canonical link.
+class DeAmpThrottle : public sniffer::SnifferThrottle {
+ public:
+  explicit DeAmpThrottle(scoped_refptr<base::SequencedTaskRunner> task_runner,
+                         DeAmpService* service,
+                         content::WebContents* contents);
+  ~DeAmpThrottle() override;
+
+  static std::unique_ptr<DeAmpThrottle> MaybeCreateThrottleFor(
+      scoped_refptr<base::SequencedTaskRunner> task_runner,
+      DeAmpService* service,
+      content::WebContents* contents);
+
+  // Implements blink::URLLoaderThrottle.
+  void WillProcessResponse(const GURL& response_url,
+                           network::mojom::URLResponseHead* response_head,
+                           bool* defer) override;
+
+ private:
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+  DeAmpService* service_;
```

Comment:
see https://chromium.googlesource.com/chromium/src/+/master/base/memory/raw_ptr.md#when-to-use-raw_ptr_t for raw pointers